test/rate_limiter_test.cc

```diff
@@ -68,6 +68,58 @@ TEST_F(RateLimiterTest, BurstingRateLimiterTest) {
   EXPECT_FALSE(rate_limiter->tryAcquireOne());
 }
 
+TEST_F(RateLimiterTest, ScheduledStartingRateLimiterTest) {
+  Envoy::Event::SimulatedTimeSystem time_system;
+  const auto schedule_delay = 10ms;
+  // We test regular flow, but also the flow where the first aquisition attempt comes after the
+  // scheduled delay. This should be business as usual from a functional perspective, but internally
+  // this rate limiter specializes on this case to log a warning message, and we want to cover that.
+  for (const bool starting_late : std::vector<bool>{false, true}) {
+    const Envoy::MonotonicTime scheduled_starting_time =
+        time_system.monotonicTime() + schedule_delay;
+    std::unique_ptr<MockRateLimiter> mock_rate_limiter = std::make_unique<MockRateLimiter>();
+    MockRateLimiter& unsafe_mock_rate_limiter = *mock_rate_limiter;
+    InSequence s;
+
+    EXPECT_CALL(unsafe_mock_rate_limiter, timeSource)
+        .Times(AtLeast(1))
+        .WillRepeatedly(ReturnRef(time_system));
+    RateLimiterPtr rate_limiter = std::make_unique<ScheduledStartingRateLimiter>(
+        std::move(mock_rate_limiter), scheduled_starting_time);
+    EXPECT_CALL(unsafe_mock_rate_limiter, tryAcquireOne)
+        .Times(AtLeast(1))
+        .WillRepeatedly(Return(true));
+
+    if (starting_late) {
```

Comment:
I'm not sure if starting_late is adding any extra coverage for this test case.  This test should already be testing what happens when enough time has passed and when not enough time has passed.

Reply:
I'm not sure I follow; the intent here is to skip the scheduled delay entirely, so the first call to `acquireOne()` will hit the line that logs the warning that the "first acquisition attempt was late":

https://github.com/envoyproxy/nighthawk/pull/281/files/de66b5e48997d31728125385f33265bfb62094df#diff-698caf1e153397f66dc8ca94b64a391fR72

Reply:
Ah, okay. I didn't process the comments in line 74-76 properly when I read this code the first time.  I follow now.

Is there a way to assert that the log line was actually hit? Or is this just for line code coverage?

Reply:
This is just for coverage. Unfortunately we can't really mock the logging today and I couldn't find an efficient way to assert on hitting that line